hal/armv7m/stm32/l4/stm32l4.c

```diff
@@ -0,0 +1,490 @@
+	else if (d <= ahb2_end)
+		*hz = !!(*(stm32_common.rcc + rcc_ahb2enr) & (1 << (d - ahb2_begin)));
+	else if (d <= ahb3_end)
+		*hz = !!(*(stm32_common.rcc + rcc_ahb3enr) & (1 << (d - ahb3_begin)));
+	else if (d <= apb1_1_end)
+		*hz = !!(*(stm32_common.rcc + rcc_apb1enr1) & (1 << (d - apb1_1_begin)));
+	else if (d <= apb1_2_end)
+		*hz = !!(*(stm32_common.rcc + rcc_apb1enr2) & (1 << (d - apb1_2_begin)));
+	else if (d <= apb2_end)
+		*hz = !!(*(stm32_common.rcc + rcc_apb2enr) & (1 << (d - apb2_begin)));
+	else if (d == pctl_rtc)
+		*hz = !!(*(stm32_common.rcc + rcc_bdcr) & (1 << 15));
+	else
+		return -1;
+
+	return 0;
+}
+
+
+int _stm32_rccSetCPUClock(u32 hz)
+{
+	u8 range;
+	u32 t;
+
+	if (hz <= 100 * 1000) {
+		range = 0;
+		hz = 100 * 1000;
+	}
+	else if (hz <= 200 * 1000) {
+		range = 1;
+		hz = 200 * 1000;
+	}
+	else if (hz <= 400 * 1000) {
+		range = 2;
+		hz = 400 * 1000;
+	}
+	else if (hz <= 800 * 1000) {
+		range = 3;
+		hz = 800 * 1000;
+	}
+	else if (hz <= 1000 * 1000) {
+		range = 4;
+		hz = 1000 * 1000;
+	}
+	else if (hz <= 2000 * 1000) {
+		range = 5;
+		hz = 2000 * 1000;
+	}
+	else if (hz <= 4000 * 1000) {
+		range = 6;
+		hz = 4000 * 1000;
+	}
+	else if (hz <= 8000 * 1000) {
+		range = 7;
+		hz = 8000 * 1000;
+	}
+	else if (hz <= 16000 * 1000) {
+		range = 8;
+		hz = 16000 * 1000;
+	}
+	/* TODO - we need to change flash wait states to handle below frequencies
+	else if (hz <= 24000 * 1000) {
+		range = 9;
+		hz = 24000 * 1000;
+	}
+	else if (hz <= 32000 * 1000) {
+		range = 10;
+		hz = 32000 * 1000;
+	}
+	else if (hz <= 48000 * 1000) {
+		range = 11;
+		hz = 48000 * 1000;
+	}
+	*/
+	else {
+		/* We can use HSI, if higher frequency is needed */
+		return -1;
+	}
+
+	if (hz > 6000 * 1000)
+		_stm32_pwrSetCPUVolt(1);
+
+	while (!(*(stm32_common.rcc + rcc_cr) & 2))
+		;
+
+	t = *(stm32_common.rcc + rcc_cr) & ~(0xf << 4);
+	*(stm32_common.rcc + rcc_cr) = t | range << 4 | (1 << 3);
+	hal_cpuDataMemoryBarrier();
+
+	if (hz <= 6000 * 1000)
+		_stm32_pwrSetCPUVolt(2);
+
+	stm32_common.cpuclk = hz;
+
+	return 0;
+}
+
+
+u32 _stm32_rccGetCPUClock(void)
+{
+	return stm32_common.cpuclk;
+}
+
+
+void _stm32_rccClearResetFlags(void)
+{
+	*(stm32_common.rcc + rcc_csr) |= 1 << 23;
+}
+
+
+/* RTC */
+
+
+void _stm32_rtcUnlockRegs(void)
+{
+	/* Set DBP bit */
+	*(stm32_common.pwr + pwr_cr1) |= 1 << 8;
+
+	/* Unlock RTC */
+	*(stm32_common.rtc + rtc_wpr) = 0x000000ca;
+	*(stm32_common.rtc + rtc_wpr) = 0x00000053;
+}
+
+
+void _stm32_rtcLockRegs(void)
+{
+	/* Lock RTC */
+	*(stm32_common.rtc + rtc_wpr) = 0x000000ff;
+
+	/* Reset DBP bit */
+	*(stm32_common.pwr + pwr_cr1) &= ~(1 << 8);
+}
+
+
+/* PWR */
+
+
+void _stm32_pwrSetCPUVolt(u8 range)
+{
+	u32 t;
+
+	if (range != 1 && range != 2)
+		return;
+
+	t = *(stm32_common.pwr + pwr_cr1) & ~(3 << 9);
+	*(stm32_common.pwr + pwr_cr1) = t | (range << 9);
+
+	/* Wait for VOSF flag */
+	while (*(stm32_common.pwr + pwr_sr2) & (1 << 10))
+		;
+}
+
+
+/* SysTick */
+
+
+int _stm32_systickInit(u32 interval)
```

Comment:
What do you think about moving systick to `cpu` module? It is a generic part for Cortex-M as well. Then You can remove `scb` from stm32.

Reply:
I cannot remove it, if I want to enable exceptions: https://github.com/phoenix-rtos/plo/blob/ab4d7e29a3acef43f4a7375f9b67d414b990b5e3/hal/armv7m/stm32/l4/stm32l4.c#L489

Reply:
Look at the comment below.

Reply:
Can't really do that, systick clock config is MCU integrator dependent

Reply:
Ok